tractor/trionics/_supervisor.py

```diff
@@ -0,0 +1,256 @@
+# tractor: structured concurrent "actors".
+# Copyright 2018-eternity Tyler Goodlet.
+
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+'''
+Erlang-style (ish) "one-cancels-one" nursery.
+
+'''
+from __future__ import annotations
+from contextlib import (
+    asynccontextmanager as acm,
+    contextmanager as cm,
+    nullcontext,
+)
+from typing import ContextManager
+
+from outcome import (
+    Outcome,
+    acapture,
+)
+import pdbp
+from msgspec import Struct
+import trio
+from trio._core._run import (
+    Task,
+    CancelScope,
+    Nursery,
+)
+
+class MaybeOutcome(Struct):
+
+    _ready: Event = trio.Event()
+    _outcome: Outcome | None = None
+    _result: Any | None = None
+
+    @property
+    def result(self) -> Any:
+        '''
+        Either Any or None depending on whether the Outcome has compeleted.
+
+        '''
+        if self._outcome is None:
+            raise RuntimeError(
+                # f'Task {task.name} is not complete.\n'
+                f'Outcome is not complete.\n'
+                'wait on `await MaybeOutcome.unwrap()` first!'
+            )
+        return self._result
+
+    def _set_outcome(
+        self,
+        outcome: Outcome,
+    ):
+        self._outcome = outcome
+        self._result = outcome.unwrap()
+        self._ready.set()
+
+    # TODO: maybe a better name like,
+    # - .wait_and_unwrap()
+    # - .wait_unwrap()
+    # - .aunwrap() ?
+    async def unwrap(self) -> Any:
+        if self._ready.is_set():
+            return self._result
+
+        await self._ready.wait()
+
+        out = self._outcome
+        if out is None:
+            raise ValueError(f'{out} is not an outcome!?')
+
+        return self.result
+
+
+class TaskHandle(Struct):
+    task: Task
+    cs: CancelScope
+    exited: Event | None = None
+    _outcome: Outcome | None = None
+
+
+class ScopePerTaskNursery(Struct):
+    _n: Nursery
+    _scopes: dict[
+        Task,
+        tuple[CancelScope, Outcome]
+    ] = {}
+
+    scope_manager: ContextManager | None = None
+
+    async def start_soon(
```

Comment:
I don't really know what the best API would be for doing this with `.start()`?

1. you don't really need start since you can always implement it by getting the child task to `Event.set()`, or even further use this `MaybeOutcome` thing from above.
2. how do you get the `cancel_scope_manager` yielded output in that case, since `.start()` would normally return whatever the child passed to `status.started()`?
  - i.e. you can't do `started = await n.start(blah)` and expect `started` to be the `outcome, cs` stuff
3. maybe the child task could get access to it's own cancel scope "stuff" from somewhere and then choose to `status.started((outcome, cs))` it back to the caller?

Reply:
Huh, i guess we could just offer some kinda support for hooking into the started value returned by the task? that way the `cancel_scope_manager` could decide what to send back to the `.start()` caller with the default being the value that the child `.started()` ?

Reply:
huh, likely you could implement that as a double yield generator then?

I dunno what would be best for ensuring that other then decorator params and then `async_fn` checking inside `.start()` against that i guess?